Assets/HoloToolkit/Utilities/Scripts/Solvers/Solver.cs

```diff
@@ -0,0 +1,411 @@
+//
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+//
+using UnityEngine;
+using System.Collections;
+#if UNITY_2017_2_OR_NEWER
+using UnityEngine.XR.WSA.Input;
+#else
+using UnityEngine.VR.WSA.Input;
+#endif
+
+namespace HoloToolkit.Unity
+{
+    [RequireComponent(typeof(SolverHandler))]
+
+    /// <summary>
+    ///   SolverBase is the base abstract class for all Solvers to derive from.  It provides state tracking, smoothing parameters
+    ///   and implementation, automatic solver system integration, and update order.  Solvers may be used without a link,
+    ///   as long as UpdateLinkedTransform is false.
+    /// </summary>
+    public abstract class Solver : MonoBehaviour
+    {
+
+        #region public members
+        [Tooltip("If true, the position and orientation will be calculated, but not applied, for other components to use")]
+        public bool UpdateLinkedTransform = false;
+
+        [Tooltip("Position lerp multiplier")]
+        public float MoveLerpTime = 0.1f;
+        [Tooltip("Rotation lerp multiplier")]
+        public float RotateLerpTime = 0.1f;
+        [Tooltip("Scale lerp multiplier")]
+        public float ScaleLerpTime = 0;
+
+        [Tooltip("If true, the Solver will respect the object's original scale values")]
+        public bool MaintainScale = true;
+
+        // Field that may be used to represent final position to be smoothly attained
+        [HideInInspector]
+        public Vector3 GoalPosition;
+        // Field that may be used to represent final rotation to be smoothly attained
+        [HideInInspector]
+        public Quaternion GoalRotation;
+        // Field that may be used to represent final scale to be smoothly attained
+        [HideInInspector]
+        public Vector3 GoalScale;
+
+        [Tooltip("Working output is smoothed if true.  Otherwise snapped")]
+        public bool Smoothing = true;
+
+        [Tooltip("If > 0, this solver will deactivate after this much time, even if the state is still active")]
+        public float Lifetime = 0;
+        #endregion
+
+        #region private members
+        protected SolverHandler solverHandler;
+        protected SolverHandler.TrackedObjectToReferenceEnum oldHandler;
```

Comment:
Should `TracktedObjectToReferenceEnum`  be outside the scope of `SolverHandler`?

Reply:
I went back and forth on this. The reason its in SolverHandler is to prevent duplicate TrackedObjectToReference fields. Since there can only be one SolverHandler component on an object, there is no worry of having TrackedObjectToReference being set multiple times - which would be the case if its in Solver (where it was originally). While this isn't the prettiest scope when extending Solver, it makes for an extremely clear editor component experience.